### DISCUSSION TITLE
Separating segments of executables

### DISCUSSION THREAD

**Discussion by astevins** [sid-agrawal/OSmosis#20] [Knowledge Base]
For scenarios where we want to create two address spaces with some shared and some separate regions, we may need to put some code/data into a separate segment of the ELF, so that it can be loaded to just one address space and not another. For example, we may want to send a library's data section and an app's data section to separate segments. 

Initially, for the SQLite example, I added a small linker file which is used in addition to the default linker script:
```
SECTIONS
{
  .lib_data :
  {
    apps/sel4test-driver/sel4test-tests/sqlite3/libsqlite3.a(.data)
    apps/sel4test-driver/kvstore/libkvstore_server_lib.a(.data)
  }
}
INSERT AFTER .data;
```

This sends the library data to a separate section, but the default linker still places the section within the greater data segment, so it cannot be handled separately during elf loading. 

After some investigation, it appears that it is not possible to create a new segment while still using the default linker script. It should be possible to use an entirely custom linker script to generate the desired ELF instead. 
Reference provided by Reto: https://github.com/achreto/arm-fastmodels-boot/blob/main/src/boot.lds.in

In a high level, I imagine the script would look something like this:
```
PHDRS 
{
  data PT_LOAD;
  app_data PT_LOAD;
  lib_data PT_LOAD;
  text PT_LOAD;
  <debug?>
}

SECTIONS
{
  .lib_data :
  {
    apps/sel4test-driver/sel4test-tests/sqlite3/libsqlite3.a(.data)
    apps/sel4test-driver/kvstore/libkvstore_server_lib.a(.data)
  } : lib_data

  .app_data :
  {
    <app data files>
  } : app_data

  .data :
  {
    <shared data files>
  } : data

  .text :
  {
    <all code>
  } : text

  ...
}
```

The script would need more than this, I'm not certain which parts are needed from the [default linker script](https://github.com/user-attachments/files/16184411/aarch64-none-elf-default-linker.txt)

**Comment by astevins** [comment]
I will also document what we did previously for the HighJmp example:
- The app and lib have completely separate heaps. This covers almost all (if not all) SQLite data, since SQLite uses the heap for its data.
- We deep-copy the data segment, so both the app and lib have a separate copy of all data (so they cannot see anything written to it by eachother, but they will see whatever the initial values from the ELF were). This is where, in a more ideal scenario, the data segment would be split into multiple segments, the app would only have its own data segment loaded, etc .
- The code segment is shared

**Reply by p-linh** [reply]
Additional note about the separate heaps. Since we have not split the data segment into more granular slices, the musllibc data (containing `malloc`'s bookkeeping metadata) are replicated between the app and lib, which requires that the heap be deep copied between the app and lib. 

The lib would technically have access to whatever was in the heap prior to deep-copying. Due to separation of the heap, at the very least, musllibc data must be separated as `malloc` metadata will diverge between the app and lib.

Workarounds that can be employed:

- simply not using the heap until the second ADS has been set up by the app
- resetting musllibc data to its original values (from loading in the ELF), so the lib's heap can be set up completely empty

**Comment by sid-agrawal** [comment]
Could be a good learning resource: 
https://home.cs.colorado.edu/~main/cs1300/doc/gnu/ld_3.html
